src/rhsmlib/cloud/utils.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2020 Red Hat, Inc.
+#
+# This software is licensed to you under the GNU General Public License,
+# version 2 (GPLv2). There is NO WARRANTY for this software, express or
+# implied, including the implied warranties of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+# along with this software; if not, see
+# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+#
+# Red Hat trademarks are not licensed under GPLv2. No permission is
+# granted to use or replicate Red Hat trademarks that are incorporated
+# in this software or its documentation.
+#
+
+"""
+This module contains several utils used for VMs running on clouds
+"""
+
+import logging
+
+from rhsmlib.facts.host_collector import HostCollector
+from rhsmlib.facts.hwprobe import HardwareCollector
+
+from rhsmlib.cloud.providers.aws import AWSCloudDetector
+from rhsmlib.cloud.providers.azure import AzureCloudDetector
+from rhsmlib.cloud.providers.gcp import GCPCloudDetector
+
+# List of classes with supported cloud providers
+CLOUD_DETECTORS = [
+    AWSCloudDetector,
+    AzureCloudDetector,
+    GCPCloudDetector
+]
+
+log = logging.getLogger(__name__)
+
+
+def detect_cloud_provider():
+    """
+    This method tries to detect cloud provider using hardware information provided by dmidecode.
+    When there is strong sign that the VM is running on one of the cloud provider, then return
+    list containing only one provider. When there is no strong sign of one cloud provider, then
+    try to detect cloud provider using heuristics methods. In this case this method will return
+    list of all cloud providers sorted according detected probability
+    :return: List of string representing detected cloud providers. E.g. ['aws'] or ['aws', 'gcp']
+    """
+
+    # Gather only information about hardware and virtualization
+    facts = {}
+    facts.update(HostCollector().get_all())
+    facts.update(HardwareCollector().get_all())
+
+    cloud_detectors = [cls(facts) for cls in CLOUD_DETECTORS]
+
+    log.debug('Trying to detect cloud provider')
+
+    # First try to detect cloud providers using strong signs
+    for cloud_detector in cloud_detectors:
+        cloud_detected = cloud_detector.is_running_on_cloud()
+        if cloud_detected is True:
+            return [cloud_detector.ID]
```

Comment:
How sure are we that each of the facts checked in each of the cloud detectors is_running_on_cloud methods are mutually exclusive?

For example, can be sure there are no vms with a dmi.chassis.asset_tag of "7783-7084-3265-9085-8269-3286-77" and also "google" in the dmi.bios.vendor?
 (might be nice to have a unittest for this to solidify what our expectations are in this case).

If we can't be sure that these indicators are mutually exclusive I'd suggest we rely on the heuristic below, combined with taking the result of whichever CloudMetaDataProvider returns a result (are the methods of gathering the metadata mutually exclusive? My intuition says yes but I've yet to find a concrete answer to this).

Reply:
> How sure are we that each of the facts checked in each of the cloud detectors is_running_on_cloud methods are mutually exclusive?

I did several experiments and I can say that these facts are mutually exclusive ATM I don't believe that e.g. Azure will add any virtual device including Google keyword.

> For example, can be sure there are no vms with a dmi.chassis.asset_tag of "7783-7084-3265-9085-8269-3286-77" and also "google" in the dmi.bios.vendor?

Probability of such case is IMHO very low, because dmi.chassis.asset_tag is not set at all on GCP.

> (might be nice to have a unittest for this to solidify what our expectations are in this case).
> 
> If we can't be sure that these indicators are mutually exclusive I'd suggest we rely on the heuristic below, combined with taking the result of whichever CloudMetaDataProvider returns a result (are the methods of gathering the metadata mutually exclusive? My intuition says yes but I've yet to find a concrete answer to this).

I would love to see some specification here: https://www.dmtf.org/ for VMs running on public clouds, but there is nothing there and Microsoft/Google/Amazon/IBM are not even members of DMTF. In theory they could set some OEM strings using this approach: https://libvirt.org/formatdomain.html#smbios-system-information

Fragment of VM xml file:
```xml
  <sysinfo type="smbios">
    <oemStrings>
      <entry>public_cloud:id aws</entry>
    </oemStrings>
  </sysinfo>
  <os>
    <type arch="x86_64" machine="pc-q35-4.2">hvm</type>
    <boot dev="hd"/>
    <smbios mode="sysinfo"/>
  </os>
```

I'm able to see the OEM string in output of dmidecode:

```
Handle 0x0E00, DMI type 11, 5 bytes
OEM Strings
	String 1: public_cloud:id aws
```

Can we convince cloud providers to do such thing? It would be great, but I guess that we don't have power to do it.

Reply:
I guess what I'm trying to say is, if we aren't sure that the results returned by "is_running_on_cloud" are accurate, allowing that methods return value to determine the entirety of the list of what metadata providers we check could result in missed detections.

Let's say that my case from the last comment were true and that this system is actually running on GCP. In that case this method would return only Azure. This will result in a failure to retrieve metadata from the correct provider.

Reply:
So, do you propose to completely remove the method `is_running_on_cloud` and use only `is_likely_running_on_cloud`?

Reply:
I think that would be more robust of a solution in my opinion. Or we could first use the ```is_running_on_cloud``` as the first check for inclusion in the list. Then add to that list with the ```is_likely_running_on_cloud```.

Reply:
Yeah, we use this approach. We use `is_running_on_cloud()` and if nothing is found, then we use `is_likely_running_on_cloud()`. It is used here: https://github.com/candlepin/subscription-manager/blob/jhnidek/ent-3288_detect_cloud_provider/src/rhsmlib/cloud/utils.py#L40